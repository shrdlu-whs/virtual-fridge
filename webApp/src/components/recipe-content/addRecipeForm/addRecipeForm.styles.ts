import { makeStyles } from "@material-ui/core";

export const useStyles = makeStyles((theme) => ({
  addRecipeContentContainer: {
    padding: theme.spacing(5, 5),
  },
  
  addRecipeTitle: {
    marginTop: 0,
    textAlign: "center",
    marginBottom: theme.spacing(3),
    textTransform: "uppercase",
    paddingBottom: theme.spacing(2),
    borderBottom: `1px solid ${theme.palette.primary.main}`,
  },
}));
