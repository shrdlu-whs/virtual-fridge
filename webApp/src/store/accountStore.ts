/* eslint-disable no-debugger */
import { makeObservable, action, computed, observable } from "mobx";
import { Auth } from "aws-amplify";
import { RootStore } from "./rootStore";
import {
  IRegisterForm,
  ISignInForm,
  IForgotPasswordForm,
  IForgotPasswordSubmitForm,
  IUserInfo,
} from "../interfaces/account.interface";
import axios, { Canceler, CancelToken, CancelTokenSource, CancelTokenStatic } from "axios";


// Amplify: https://docs.amplify.aws/lib/auth/getting-started/q/platform/js

export class AccountStore {
  @observable userInfo: IUserInfo | null = null;
  @observable forgotUsername: string | undefined = undefined;

  // cancelToken: CancelTokenSource = axios.CancelToken.source()
  cancelRequest: Canceler | undefined = undefined;

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  generateCancelToken() {
    return new axios.CancelToken((c) => {
      this.cancelRequest = c;
    })
  }

  @action
  async register(register: IRegisterForm) {
    try {
      await Auth.signUp({
        username: register.username,
        password: register.password,
        attributes: {
          email: register.email,
        },
      });
      this.rootStore.routeStore.redirect("/signIn");
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "register",
        "error",
        true,
        error
      );
    }
  }

  @action
  async signIn(signIn: ISignInForm) {
    try {
      this.rootStore.uiStateStore.setSubmitting(true);
      await Auth.signIn(signIn.username, signIn.password);
      await this.fetchUserInfo();
      this.rootStore.routeStore.redirect("/products");
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "signIn",
        "error",
        true,
        error
      );
    } finally {
      this.rootStore.uiStateStore.setSubmitting(false);
    }
  }

  @action
  async signOut() {
    try {
      await Auth.signOut();
      // this.userInfo = null; // If the site refresh, fetchUserInfo() will be executed and the userInfo will set anyway
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "signOut",
        "error",
        true,
        error
      );
    } finally {
      this.rootStore.routeStore.redirect("/signIn");
    }
  }

  @action
  async forgotPassword({ username }: IForgotPasswordForm) {
    try {
      await Auth.forgotPassword(username);
      this.setForgotUsername(username);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "forgot_password",
        "error",
        true,
        error
      );
    }
  }

  @action
  async forgotPasswordSubmit({
    username,
    code,
    password,
  }: IForgotPasswordSubmitForm) {
    try {
      await Auth.forgotPasswordSubmit(username, code, password);
      this.rootStore.routeStore.redirect("/signIn");
      this.setForgotUsername(undefined);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "forgot_password_submit",
        "error",
        true,
        error
      );
    }
  }

  @action
  async changePassword(oldPassword: string, newPassword: string) {
    try {
      const user = await Auth.currentAuthenticatedUser();
      await Auth.changePassword(user, oldPassword, newPassword);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "change_password",
        "error",
        true,
        error
      );
    }
  }

  @action
  async fetchUserInfo() {
    if (!this.rootStore.uiStateStore.fetchingUser) {
      try {
        this.rootStore.uiStateStore.setFetchingUser(true);
        const userInfo = (await Auth.currentUserInfo()) as IUserInfo | null;
        // const userInfo: IUserInfo = {username: "Test", attributes: {email: "test@mail.com", email_verified: "true", sub: "" }};
        this.setUserInfo(userInfo);
      } catch (error) {
        this.setUserInfo(null);
        this.rootStore.notificationStore.createNotification(
          "fetch_user",
          "error",
          true,
          error
        );
      } finally {
        this.rootStore.uiStateStore.setFetchingUser(false);
        // Its initiating, also when the an error occured => the user returns to login page
        if(!this.rootStore.uiStateStore.isInitiating) {
          this.rootStore.uiStateStore.setInitiated(true);
        }
      }
    }
  }

  @action
  setUserInfo(userInfo: IUserInfo | null) {
    this.userInfo = userInfo;
  }

  @action
  setForgotUsername(username: string | undefined) {
    this.forgotUsername = username;
  }
}
