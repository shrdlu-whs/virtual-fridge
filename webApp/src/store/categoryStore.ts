/* eslint-disable no-debugger */
import { makeObservable, action, observable, runInAction } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist";
import { IApiCategory } from "../interfaces/api/api-category.interface";
import { getCategories } from "../services/category2.service";

export class CategoryStore {
  @observable categories: IApiCategory[] = [];

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  async getCategories() {
    try {
      const categories = (await getCategories()).category;
      this.setCategories(categories);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_categories", "error", true, error);
      }
    }
  }

  @action
  setCategories(categories: IApiCategory[]) {
    this.categories = categories;
  }
}
