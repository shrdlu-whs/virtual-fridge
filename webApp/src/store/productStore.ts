/* eslint-disable no-debugger */
import { makeObservable, action, observable, runInAction } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist";
import { IApiProduct } from "../interfaces/api/api-product.interface";
import { IIronStockProduct } from "../interfaces/ironStockProduct.interface";
import { getProducts, getProduct, addProduct, updateProduct, deleteProduct } from "../services/product2.service";
import { IApiCategory } from "../interfaces/api/api-category.interface";
import { INutritionValue, IProductItem, IProduct } from "../interfaces/froms/addProductForm.interface";
import { EMPTY_OBJECT } from "mobx/dist/internal";

export class ProductStore {

  nutritionValueEmpty: INutritionValue = {
    calories: 0,
    carbohydrates: 0,
    fat: 0,
    fiber: 0,
    nutritionScore: "A",
    protein: 0,
    salt: 0,
    saturatedFat: 0,
    sugar: 0
  };

  @observable activeProduct: IApiProduct | undefined = undefined;
  @observable products: IApiProduct[] = [];
  @observable ironStockProducts: IIronStockProduct[] = [];
  @observable missingIronStockProducts: IIronStockProduct[] = [];
  @observable productItems: IProductItem[] = [];

  @observable addProductFormData: IProduct = {
    name: "",
    iconPath: "",
    category: [],
    totalQuantity: 1,
    minQuantity: -1,
    unit: "",
    items: this.productItems,
    nutritionalValue: this.nutritionValueEmpty
  };

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  async getProducts() {
    try {
      // const products = (await getProducts(this.rootStore.accountStore.generateCancelToken())).data;
      this.rootStore.uiStateStore.setLoading(true);
      const products = (await getProducts()).products;
      this.setProducts(products);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_products", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async getProduct(id: number) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      // const product = (await getProduct(id, this.rootStore.accountStore.generateCancelToken())).data;
      const product = (await getProduct(id)).product;
      this.setActiveProduct(product);
      this.updateCachedProducts(product);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_product", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async addProduct(product: IApiProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      product = (await addProduct(product)).createProduct;
      this.setProducts(this.products.concat([product]));
    } catch (error) {
      console.log(error)
      this.rootStore.notificationStore.createNotification("add_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async updateProduct(product: IApiProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      await updateProduct(product);
      if (product.id) this.updateCachedProducts(product);
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async deleteProduct(id: number) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      await deleteProduct(id);
      const index = this.products.findIndex((product) => product.id === id);
      if (index !== -1) {
        runInAction(() => {
          this.products.splice(index, 1);
        });
      }
    } catch (error) {
      this.rootStore.notificationStore.createNotification("delete_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  @action
  updateCachedProducts(product: IApiProduct) {
    // Update cached products
    const index = this.products.findIndex((sProduct) => sProduct.id === product.id);
    if (index !== -1) {
      this.products.splice(index, 1, product);
    } else {
      this.products.push(product);
    }
  }

  async getIronStockProducts() {
    try {
      const products = (await getProducts()).products;
      this.rootStore.uiStateStore.setLoading(true);
      this.setProducts(products);
      const ironStockProducts = products.map(
        (product): IIronStockProduct => {
          return { id: product.id!, name: product.name, stock: product.totalQuantity, ironStock: product.minQuantity };
        }
      );
      this.setIronStockProducts(ironStockProducts.filter((product) => product.ironStock >= 0));
      this.setMissingIronStockProducts(ironStockProducts.filter((product) => product.ironStock < 0));
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_products", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async addIronStockProduct(ironStockProduct: IIronStockProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      const cachedProduct = this.products.find((sProduct) => sProduct.id === ironStockProduct.id)!;
      cachedProduct.minQuantity = ironStockProduct.ironStock;
      await updateProduct(cachedProduct);
      this.setIronStockProducts(this.ironStockProducts.concat([ironStockProduct]));
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async updateIronStockProducts(ids: number[]) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      for (const id of ids) {
        const product = this.ironStockProducts.find((product) => product.id === id);
        if (product) {
          const cachedProduct = this.products.find((sProduct) => sProduct.id === product.id)!;
          cachedProduct.minQuantity = product.ironStock;
          await updateProduct(cachedProduct);
        }
      }
    } catch (error) {
      // Rollback ???
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  @action
  updateCachedIronStockProduct(id: number, ironStock: number) {
    const ironStockProduct = this.ironStockProducts.find((product) => product.id === id);
    if (ironStockProduct) {
      ironStockProduct.ironStock = ironStock;
    }
  }

  @action
  async deleteIronStockProduct(id: number) {
    try {
      const cachedProduct = this.products.find((sProduct) => sProduct.id === id)!;
      cachedProduct.minQuantity = -1;
      await updateProduct(cachedProduct);
      this.setIronStockProducts(this.ironStockProducts.filter((product) => product.id !== id));
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    }
  }

  @action
  setActiveProduct(product: IApiProduct) {
    this.activeProduct = product;
  }

  @action
  setProducts(products: IApiProduct[]) {
    this.products = products;
  }

  @action
  setIronStockProducts(ironStockProducts: IIronStockProduct[]) {
    this.ironStockProducts = ironStockProducts;
  }

  @action
  setMissingIronStockProducts(ironStockProducts: IIronStockProduct[]) {
    this.missingIronStockProducts = ironStockProducts;
  }

  @action
  setProductItems(items: IProductItem[]) {
    this.productItems = items;
  }

  @action
  setAddProductFormData(data: IProduct) {
    this.addProductFormData = data;
  }

  @action
  resetAddProductFormData = () => {
    this.setProductItems([]);
    this.setAddProductFormData({
      name: "",
      iconPath: "",
      category: [],
      totalQuantity: 1,
      minQuantity: -1,
      unit: "",
      items: this.productItems,
      nutritionalValue: this.nutritionValueEmpty
    });
  };
}
