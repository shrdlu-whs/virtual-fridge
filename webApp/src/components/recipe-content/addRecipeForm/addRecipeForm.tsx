import React, { useEffect } from "react";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";
import InputLabel from "@material-ui/core/InputLabel";
import Select from "@material-ui/core/Select";
import { useTranslation } from "react-i18next";
import { IIronStockProduct } from "../../../interfaces/ironStockProduct.interface";
import { Formik } from "formik";
import * as Yup from "yup";
import { v4 } from "uuid";
import RequestButton from "../../request-button/requestButton";
import ExitToApp from "@material-ui/icons/ExitToApp";
import { IApiRecipe } from "../../../interfaces/api/api-recipe.interface";
import { IApiRecipeItem } from "../../../interfaces/api/api-recipeItem.interface";
import { IRecipeItem, IRecipe } from "../../../interfaces/froms/addRecipeForm.interface";
import { FormControl, Grid, Paper, Toolbar, Typography } from "@material-ui/core";
import { useStyles } from "./addRecipeForm.styles";
import { getMillesFromString } from "../../../utils/time.utils";
import IngredientsTable from "./ingredientsTable/ingredients";
import { ProductStore } from "../../../store/productStore";
import { RecipeStore } from "../../../store/recipeStore";
import { UiStateStore } from "../../../store/uiStateStore";
import { AddIngredientForm } from "./addIngredientForm/addIngredientForm";
import { inject, observer } from "mobx-react";
import { runInAction } from "mobx";
import { TextareaAutosize } from "@material-ui/core";

interface IAddRecipeDialogProps {
  recipeStore?: RecipeStore;
  productStore?: ProductStore;
  uiStateStore?: UiStateStore;
  isSubmitting: boolean;
  isMobile: boolean;
  handleAddRecipe: (recipe: IApiRecipe) => void;
  handleAddRecipeCancel: () => void;
}

export const AddRecipeDialog = ({
  productStore,
  recipeStore,
  uiStateStore,
  isSubmitting,
  isMobile,
  handleAddRecipe,
  handleAddRecipeCancel,
}: IAddRecipeDialogProps) => {
  const { t } = useTranslation();
  const classes = useStyles();

  const difficulties = [
    { step: t(`value.recipe_difficulty_easy`), id: "easy" },
    { step: t(`value.recipe_difficulty_intermediate`), id: "intermediate" },
    { step: t(`value.recipe_difficulty_difficult`), id: "difficult" },
  ];

  React.useEffect(() => {
    productStore!.getProducts();
  }, []);

  const validateRecipe = (formData: IRecipe) => {
    const recipe: IApiRecipe = {
      name: formData.name,
      shortDescription: formData.shortDescription,
      instructions: formData.instructions,
      hyperlink: formData.hyperlink,
      expectedTime: getMillesFromString(formData.expectedTime),
      difficulty: formData.difficulty,
      // category: [],
      recipeItems: formData.recipeItems.map(
        (item): IApiRecipeItem => {
          return { quantity: item.quantity, product: item.product };
        }
      ),
      favorite: false,
      iconPath: "",
    };
    handleAddRecipe(recipe);
    recipeStore!.resetAddRecipeFormData();
  };

  const handleAddRecipeCancelClick = () => {
    recipeStore!.resetAddRecipeFormData();
    handleAddRecipeCancel();
  };

  const handleAddIngredient = (recipeItem: IRecipeItem) => {
    runInAction(() => recipeStore!.addRecipeFormData.recipeItems.push(recipeItem));
  };

  const handleRemoveIngredients = (uuids: string[]) => {
    for (const uuid of uuids) {
      const index = recipeStore!.addRecipeFormData.recipeItems.findIndex((ingredient) => ingredient.uuid === uuid);
      if (index !== -1) {
        runInAction(() => recipeStore!.addRecipeFormData.recipeItems.splice(index, 1));
      }
    }
  };

  const handleUpdateIngredient = (ingredient: IRecipeItem, quantity: number) => {
    runInAction(() => (ingredient.quantity = quantity));
  };

  return (
    <Grid container direction="row" justify="center" alignItems="center">
      <Grid item sm={9} md={5} lg={4} xl={3}>
        <Paper className={classes.addRecipeContentContainer}>
          <Typography className={classes.addRecipeTitle} component="h1" variant="h5">
            {t("title.recipe_add_recipe_header")}
          </Typography>
          <Formik
            initialValues={recipeStore!.addRecipeFormData}
            onSubmit={(values) => {
              validateRecipe(values);
            }}
            validationSchema={Yup.object().shape({
              hyperlink: Yup.string().url().required(t("label.no_valid_hyperlink")),
              shortDescription: Yup.string().max(255, t("label.no_valid_varChar255")),
            })}
            enableReinitialize
          >
            {({ values, touched, errors, dirty, handleChange, handleBlur, handleSubmit, isValid }) => {
              return (
                <form onSubmit={handleSubmit}>
                  <div>
                    <TextField
                      label={t("label.recipe_name")}
                      name="name"
                      type="text"
                      variant="standard"
                      value={values.name}
                      onChange={handleChange}
                      onBlur={handleBlur}
                      helperText={touched.name && errors.name}
                      error={touched.name && errors.name !== undefined}
                      margin="normal"
                      fullWidth
                      required
                    />
                    <TextField
                      label={t("label.recipe_short_description")}
                      name="shortDescription"
                      type="text"
                      variant="standard"
                      value={values.shortDescription}
                      onChange={handleChange}
                      onBlur={handleBlur}
                      helperText={touched.shortDescription && errors.shortDescription}
                      error={touched.shortDescription && errors.shortDescription !== undefined}
                      margin="normal"
                      fullWidth
                    />
                    <TextField
                      label={t("label.recipe_instruction")}
                      name="instructions"
                      variant="standard"
                      value={values.instructions}
                      onChange={handleChange}
                      onBlur={handleBlur}
                      helperText={touched.instructions && errors.instructions}
                      error={touched.instructions && errors.instructions !== undefined}
                      margin="normal"
                      multiline
                      fullWidth
                    />
                    <Grid container spacing={2}>
                      <Grid item xs={6}>
                        <TextField
                          label={t("label.recipe_expected_time")}
                          name="expectedTime"
                          type="time"
                          value={values.expectedTime}
                          onChange={handleChange}
                          onBlur={handleBlur}
                          helperText={touched.expectedTime && errors.expectedTime}
                          error={touched.expectedTime && errors.expectedTime !== undefined}
                          InputLabelProps={{
                            shrink: true,
                          }}
                          inputProps={{
                            step: 15,
                          }}
                          margin="normal"
                          fullWidth
                        />
                      </Grid>
                      <Grid item xs={6}>
                        <FormControl margin="normal" fullWidth>
                          <InputLabel htmlFor="id">{t("label.recipe-difficulty")}</InputLabel>
                          <Select
                            native
                            name="difficulty"
                            onChange={handleChange}
                            inputProps={{
                              step: "step",
                              id: "id",
                            }}
                          >
                            {difficulties.map((item) => (
                              <option key={item.id} value={item.id}>
                                {item.step}
                              </option>
                            ))}
                          </Select>
                        </FormControl>
                      </Grid>
                    </Grid>
                    <IngredientsTable
                      handleDeleteIngredient={handleRemoveIngredients}
                      handleUpdateIngredient={handleUpdateIngredient}
                      isMobile={isMobile}
                      rows={values.recipeItems}
                    />
                    <AddIngredientForm products={productStore!.products} handleAddIngredient={handleAddIngredient} />
                  </div>
                  <Grid container spacing={2}>
                    <Grid item xs={6}>
                      <Button onClick={handleAddRecipeCancelClick} color="primary">
                        {t("action.cancel")}
                      </Button>
                    </Grid>
                    <Grid item xs={6}>
                      <RequestButton
                        variant="contained"
                        color="primary"
                        fullWidth
                        type="submit"
                        loading={isSubmitting}
                        disabled={!touched || isSubmitting || (dirty && !isValid)}
                        endIcon={<ExitToApp />}
                      >
                        {t("action.add")}
                      </RequestButton>
                    </Grid>
                  </Grid>
                </form>
              );
            }}
          </Formik>
        </Paper>
      </Grid>
    </Grid>
  );
};

export default inject("productStore", "recipeStore", "uiStateStore")(observer(AddRecipeDialog));
