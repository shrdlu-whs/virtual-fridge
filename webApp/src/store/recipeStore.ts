/* eslint-disable no-debugger */
import { makeObservable, action, observable, runInAction } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist";
import { getRecipe, getRecipes, addRecipe, updateRecipe, deleteRecipe } from "../services/recipe2.service";
import { IApiRecipe } from "../interfaces/api/api-recipe.interface";
import { IApiItem } from "../interfaces/api/api-item.interface";
import { IApiRecipeItem } from "../interfaces/api/api-recipeItem.interface";
import { IRecipe } from "../interfaces/froms/addRecipeForm.interface";
import { IRecipeItem } from "../interfaces/froms/addRecipeForm.interface";
export class RecipeStore {
  @observable activeRecipe: IApiRecipe | undefined = undefined;
  @observable recipes: IApiRecipe[] = [];

  //#region Add recipe form data props
  @observable recipeItems: IRecipeItem[] = [];
  @observable addRecipeFormData: IRecipe = {
    name: "",
    shortDescription: "",
    instructions: "",
    hyperlink: "https://www.someimage.net",
    expectedTime: "00:05:00",
    difficulty: "easy",
    recipeItems: this.recipeItems,
  };
  //#endregion

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  async getRecipes() {
    try {
      const recipes = (await getRecipes()).recipes;
      this.setRecipes(recipes);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_recipes", "error", true, error);
      }
    }
  }

  async getRecipe(id: number) {
    try {
      const recipe = (await getRecipe(id)).recipe;
      this.setActiveRecipe(recipe);
      this.updateCachedRecipes(recipe);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_recipe", "error", true, error);
      }
    }
  }

  async addRecipe(recipe: IApiRecipe) {
    try {
      recipe = (await addRecipe(recipe)).createRecipe;
      this.setRecipes(this.recipes.concat([recipe]));
    } catch (error) {
      this.rootStore.notificationStore.createNotification("add_recipe", "error", true, error);
    }
  }

  async updateRecipe(recipe: IApiRecipe) {
    try {
      await updateRecipe(recipe);
      if (recipe.id) this.updateCachedRecipes(recipe);
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_recipe", "error", true, error);
    }
  }

  async deleteRecipe(id: number) {
    try {
      await deleteRecipe(id);
      const index = this.recipes.findIndex((product) => product.id === id);
      if (index !== -1) {
        runInAction(() => {
          this.recipes.splice(index, 1);
        });
      }
    } catch (error) {
      this.rootStore.notificationStore.createNotification("delete_recipe", "error", true, error);
    }
  }

  updateCachedRecipes(recipe: IApiRecipe) {
    // Update cached products
    const index = this.recipes.findIndex((sRecipe) => sRecipe.id === recipe.id);
    if (index !== -1) {
      this.recipes.splice(index, 1, recipe);
    } else {
      this.recipes.push(recipe);
    }
  }

  @action
  setActiveRecipe(recipe: IApiRecipe) {
    this.activeRecipe = recipe;
  }

  @action
  setRecipes(recipes: IApiRecipe[]) {
    this.recipes = recipes;
  }

  @action
  setRecipeItems(items: IRecipeItem[]) {
    this.recipeItems = items;
  }

  @action
  setAddRecipeFormData(data: IRecipe) {
    this.addRecipeFormData = data;
  }

  @action
  resetAddRecipeFormData = () => {
    this.setRecipeItems([]);
    this.setAddRecipeFormData({
      name: "",
      shortDescription: "",
      instructions: "",
      hyperlink: "https://www.someimage.net",
      expectedTime: "00:05:00",
      difficulty: "easy",
      recipeItems: this.recipeItems,
    });
  };
}
