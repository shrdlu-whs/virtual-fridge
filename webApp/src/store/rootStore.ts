import { UiStateStore } from "./uiStateStore";
import { AccountStore } from "./accountStore";
import { RouteStore } from "./routeStore";
import { SettingStore } from "./settingStore";
import { NotificationStore } from "./notificationStore"
import { ProductStore } from "./productStore"
import { RecipeStore } from "./recipeStore"
import { ShoppingListStore } from "./shoppinglistStore"
import { CategoryStore } from "./categoryStore";


export class RootStore {
  uiStateStore: UiStateStore;
  accountStore: AccountStore;
  routeStore: RouteStore;
  notificationStore: NotificationStore;
  settingStore: SettingStore;
  productStore: ProductStore;
  recipeStore: RecipeStore;
  shoppingListStore: ShoppingListStore;
  categoryStore: CategoryStore;

  constructor() {
    this.uiStateStore = new UiStateStore(this);
    this.accountStore = new AccountStore(this);
    this.routeStore = new RouteStore(this);
    this.notificationStore = new NotificationStore(this);
    this.settingStore = new SettingStore(this);
    this.productStore = new ProductStore(this);
    this.recipeStore = new RecipeStore(this);
    this.shoppingListStore = new ShoppingListStore(this);
    this.categoryStore = new CategoryStore(this);
  }
}
