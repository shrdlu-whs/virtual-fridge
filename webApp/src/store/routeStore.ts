import { makeObservable, action, observable } from "mobx";
import { fromStream } from "mobx-utils";
import { RootStore } from "./rootStore";
import { Redirect } from "react-router-dom"

export class RouteStore {
  @observable redirectedTo: string | undefined = undefined;

  constructor(private rootStore: RootStore) {
    makeObservable(this)
  }

  @action
  redirect(path: string, tokenExpired: boolean = false) {
    if (this.redirectedTo !== path) {
      if (tokenExpired) {
        this.rootStore.notificationStore.createNotification(
          "token_expired",
          "error",
          true
        );
      }
      this.redirectedTo = path;
    }
  }

  @action
  resetRedirectedTo() {
    this.redirectedTo = undefined;
  }
}
