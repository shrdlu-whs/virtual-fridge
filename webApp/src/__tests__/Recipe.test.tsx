import React from "react";
import IronStock from '../components/iron-stock-content/ironStockContainer'
import { render, screen } from '@testing-library/react'; 
import Recipe from '../components/recipe-content/recipesContainer'
import { RootStore } from "../store/rootStore";
import { Provider } from "mobx-react";
import { IRecipe } from "../interfaces/froms/addRecipeForm.interface";
import { Difficulty, IApiRecipe } from "../interfaces/api/api-recipe.interface";

const recipe = require('../components/recipe-content/recipesContainer') 

test('render about link', () => {
    const rootStore: RootStore = new RootStore();
    render(
        <Provider {...rootStore}>
            <Recipe />
         </Provider>);
})

test('test object of shoppinglist', () => { 
    const recipe: IRecipe = {
        name: "Pommes",
        shortDescription: "",
        instructions: "",
        hyperlink: "https://www.someimage.net",
        expectedTime: "00:05:00",
        difficulty: "easy",
        recipeItems: [],
      }
      expect(recipe).toMatchSnapshot();
});

test('Recipes exists', () => {
    expect(typeof recipe).toEqual('object');
});

test('recipe handle show recipe', () => {
    expect(recipe.handleAddRecipeCancel).toBeUndefined();
});
test('recipe handle show recipe', () => {
    const rootStore: RootStore = new RootStore();
    render(
        <Provider {...rootStore}>
            <Recipe />
         </Provider>);
    const button=screen.getByRole('button', {name: 'add'})
    const input=screen.getByLabelText('search');
});
