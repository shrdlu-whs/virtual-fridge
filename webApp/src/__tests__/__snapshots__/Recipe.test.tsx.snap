// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test object of shoppinglist 1`] = `
Object {
  "category": Array [],
  "difficulty": 0,
  "expectedTime": "00:05:00",
  "hyperlink": "https://www.someimage.net",
  "instructions": "",
  "name": "Pommes",
  "recipeItems": Array [],
  "shortDescription": "",
}
`;
